### PR TITLE
OSIDB-4956: Fix check parser replaces spaces with underscores in equality check values

### PR DESCRIPTION
**Pull request by roduran-dev**
### problem 
CheckParser replaced all spaces with underscores, causing equality checks with multi-word values to fail. 

For example, "PS product is Red Hat Enterprise Linux" compared "Red_Hat_Enterprise_Linux"
  instead of "Red Hat Enterprise Linux".

  Impact: SLA exclusion policies (like Kpatch) and workflow conditions with multi-word values silently failed to match.

### solution
Modified desc2equals() to restore spaces in values for non-choices fields while preserving uppercase transformation for enum fields.

Closes OSIDB-4956

### CODE REVIEW

File: apps/sla/tests/test_models.py [hunk 1]
```diff
@@ -992,6 +992,107 @@ def test_in_condition(self, operator, component, accepted):
                 == accepted
             )
 
+        def test_kpatch_exclusion_policy_with_spaces(self):
+            """
+            Test the Kpatch SLA exclusion policy with multi-word product name
+            (regression test for OSIDB-4956)
+
+            This mirrors the actual Kpatch policy from ps-constants:
+            https://gitlab.cee.redhat.com/prodsec-dev/ps-constants/-/blob/master/data/sla_policies.yml
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: Do not put internal Gitlab links into the public code please.

File: apps/sla/tests/test_models.py [hunk 2]
```diff
@@ -992,6 +992,107 @@ def test_in_condition(self, operator, component, accepted):
...
+                affects=[fedora_affect],
+                embargoed=flaw.embargoed,
+                ps_update_stream=other_ps_update_stream.name,
+                type=Tracker.BTS2TYPE[other_ps_module.bts_name],
+            )
+
+            # Should NOT accept because product doesn't match
+            assert not policy.accepts(
+                TemporalContext(flaw=flaw, affect=fedora_affect, tracker=fedora_tracker)
+            ), "Kpatch policy should not match kpatch on non-RHEL products"
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: This is internal SLA/SLO policy which should not be publicly disclosed, use some test policy instead please.

File: apps/workflows/checks.py [hunk 3]
```diff
@@ -184,6 +182,11 @@ def choices_field(model, name):
                     or attr in self.TEXT_CHOICES_PROPERTIES
                 ):
                     value = value.upper()
+                else:
+                    # For non-choices fields, restore spaces that were replaced with underscores
+                    value = value.replace("_", " ")
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: What about this scenario:
1. original value to check is `foo bar_foo`
2. in `_parse_dict` this gets converted to `foo_bar_foo`
3. here it gets converted to `foo bar foo`
4. `foo bar_foo` != `foo bar foo`

**Review by JakubFrejlach — CHANGES_REQUESTED** [review]

**Comment by JakubFrejlach** [comment]
Also mark as technical/add changelog entry to fix the failing `Changelog check`

File: apps/workflows/checks.py (hunk 3 above)
Comment (roduran-dev) [reply apps/workflows/checks.py]: I see, I will add this other scenario :+1:

File: docs/CHANGELOG.md [hunk 4]
```diff
@@ -5,6 +5,8 @@ The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
 ## Unreleased
+### Added
+- Add validation to require PURLs (OSIDB-4810)
```
[the remark this answers is not in the document]
Comment (roduran-dev) [reply docs/CHANGELOG.md:11]: nop, that's correct and is from another PR

File: apps/sla/tests/test_models.py [hunk 5]
```diff
@@ -992,6 +992,110 @@ def test_in_condition(self, operator, component, accepted):
...
+            )
+            other_ps_update_stream = PsUpdateStreamFactory(ps_module=other_ps_module)
+            fedora_affect = AffectFactory(
+                flaw=flaw,
+                affectedness=Affect.AffectAffectedness.AFFECTED,
+                resolution=Affect.AffectResolution.DELEGATED,
+                ps_module=other_ps_module.name,
+                ps_component="kpatch",  # Same component
+                ps_update_stream=other_ps_update_stream.name,
+                impact=Impact.MODERATE,
```
[the remark this answers is not in the document]
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: I changed so it doesn't have a direct relation with prod configs.

File: apps/sla/tests/test_models.py [hunk 6]
```diff
@@ -992,6 +992,110 @@ def test_in_condition(self, operator, component, accepted):
...
+                ("Red Hat Enterprise Linux", "Red_Hat_Enterprise_Linux"),
+                ("Red_Hat_Enterprise_Linux", "Red Hat Enterprise Linux"),
+            ],
+        )
+        def test_policy_with_spaces(self, value, false_case):
+            """
+            Test the SLA exclusion policy with multi-word product name
+            (regression test for OSIDB-4956)
+
+            This mirrors and actually policy from ps-constants with spaces in the value:
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: `and actually` ?

File: apps/workflows/checks.py [hunk 7]
```diff
@@ -155,19 +159,22 @@ def has_element(instance):
...
+        original_value = original_value[total_chars:]
+        return original_value
+
+    def desc2equals(self, check_desc, original_value=None):
         """
         attribute to literal value equality check
 
-        currently only supports string attributes
-        which values do not contain any spaces
+        supports string attributes including those with spaces in values
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py:175]: I think `desc2not_equals` needs to be addressed as well

File: apps/sla/tests/test_models.py [hunk 8]
```diff
@@ -992,6 +992,110 @@ def test_in_condition(self, operator, component, accepted):
...
+                affects=[fedora_affect],
+                embargoed=flaw.embargoed,
+                ps_update_stream=other_ps_update_stream.name,
+                type=Tracker.BTS2TYPE[other_ps_module.bts_name],
+            )
+
+            # Should NOT accept because product doesn't match
+            assert not policy.accepts(
+                TemporalContext(flaw=flaw, affect=fedora_affect, tracker=fedora_tracker)
+            ), "Magic policy should not match kpatch on non-RHEL products"
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: ps_component is `magic` not `kpatch`

File: apps/workflows/checks.py [hunk 9]
```diff
@@ -84,7 +85,10 @@ def _parse_string(self, check_desc):
             self.desc2equals,
             self.desc2in,
         ]:
-            result = func(check_desc)
+            if func == self.desc2equals:
+                result = func(check_desc, original_check_desc)
+            else:
+                result = func(check_desc)
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: This seems fragile towards decorators for example and also less readable, I think better approach would be to just add `**kwargs` param to all the funcs and for the specific ones which actually need the original_check_desc you can have a named param as we currently do have in `desc2equals`. This way you don't need to perform this specific check for `desc2equals`.

File: apps/workflows/checks.py [hunk 10]
```diff
@@ -155,19 +159,22 @@ def has_element(instance):
 
                 return (doc, has_element)
 
-    def desc2equals(self, check_desc):
+    def get_original_value(self, value_after_underscore, original_value):
+        total_chars = len(value_after_underscore) * -1
+        original_value = original_value[total_chars:]
+        return original_value
+
+    def desc2equals(self, check_desc, original_value=None):
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: original value of what? Why not name it so it actually describes what it is -> original_check_desc

File: apps/workflows/checks.py [hunk 11]
```diff
@@ -155,19 +159,22 @@ def has_element(instance):
 
                 return (doc, has_element)
 
-    def desc2equals(self, check_desc):
+    def get_original_value(self, value_after_underscore, original_value):
+        total_chars = len(value_after_underscore) * -1
+        original_value = original_value[total_chars:]
+        return original_value
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: Get original value from `original_value`, of what? This is really hard to understand. What is actually happening here is that we are getting original value from the original check description.

File: apps/sla/tests/test_models.py [hunk 12]
```diff
@@ -992,6 +992,110 @@ def test_in_condition(self, operator, component, accepted):
...
+        @pytest.mark.parametrize(
+            "value, false_case",
+            [
+                ("Red Hat Enterprise Linux", "Red_Hat_Enterprise_Linux"),
+                ("Red_Hat_Enterprise_Linux", "Red Hat Enterprise Linux"),
+            ],
+        )
+        def test_policy_with_spaces(self, value, false_case):
+            """
+            Test the SLA exclusion policy with multi-word product name
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: This does not necessarily needs to be an exclusion policy, right? Such check for `Red Hat Enterprise Linux` can be used in any other policy.

File: apps/sla/tests/test_models.py [hunk 13]
```diff
@@ -992,6 +992,110 @@ def test_in_condition(self, operator, component, accepted):
                 == accepted
             )
 
+        @pytest.mark.parametrize(
+            "value, false_case",
+            [
+                ("Red Hat Enterprise Linux", "Red_Hat_Enterprise_Linux"),
+                ("Red_Hat_Enterprise_Linux", "Red Hat Enterprise Linux"),
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py:1017]: Can we add a test case with name which has some combination of spaces and underscores? -> `Red_Hat Enterprise Linux`

**Review by JakubFrejlach — COMMENTED** [review]
This is much better but there is still couple of things which needs some attention.

File: apps/sla/tests/test_models.py (hunk 12 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: yep, good catch,

File: apps/sla/tests/test_models.py (hunk 13 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py:1017]: Sure

Comment (roduran-dev) [reply apps/sla/tests/test_models.py:1017]: Added :+1:

File: apps/sla/tests/test_models.py (hunk 12 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: Fixed

File: apps/workflows/checks.py (hunk 11 above)
Comment (roduran-dev) [reply apps/workflows/checks.py]: ok, I will clarify beter.

File: apps/workflows/checks.py (hunk 7 above)
Comment (roduran-dev) [reply apps/workflows/checks.py:175]: ok, make sense

File: apps/sla/tests/test_models.py (hunk 8 above)
Comment (JakubFrejlach) [reply apps/sla/tests/test_models.py]: Still not addressed.

File: apps/sla/tests/test_models.py (hunk 6 above)
Comment (JakubFrejlach) [reply apps/sla/tests/test_models.py]: Still not addressed.

**Review by JakubFrejlach — COMMENTED** [review]
Lot of my previous comments were not addressed so I am not going to review further.

File: apps/sla/tests/test_models.py [hunk 14]
```diff
@@ -943,8 +943,12 @@ def test_conditions_multiple_flaws(self):
             [
                 ("in", "kpatch", True),
                 ("in", "coffee", False),
+                ("in", "red hat", True),
```
Comment (roduran-dev) [inline apps/sla/tests/test_models.py:946]: Just to be sure I added this edge case to the "in"/"no in" test

File: apps/workflows/checks.py (hunk 11 above)
Comment (roduran-dev) [reply apps/workflows/checks.py]: changed variables to a more descriptive name.

File: apps/workflows/checks.py (hunk 7 above)
Comment (roduran-dev) [reply apps/workflows/checks.py:175]: I also added "is not" cases to the test.

File: apps/sla/tests/test_models.py (hunk 6 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: Done :+1:

File: apps/sla/tests/test_models.py (hunk 8 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: I changed the test to be more direct and specific to the actual case. I removed kpatch and the ps_component (now its only ps product <is/is not> <Multi word description>

**Comment by roduran-dev** [comment]
> Lot of my previous comments were not addressed so I am not going to review further.

Hi @JakubFrejlach  sorry last time, I click the re review by mistake.  I just finish asserting all your comments.

File: apps/workflows/checks.py [hunk 15]
```diff
@@ -84,7 +85,7 @@ def _parse_string(self, check_desc):
             self.desc2equals,
             self.desc2in,
         ]:
-            result = func(check_desc)
+            result = func(check_desc, original_check_desc=original_check_desc)
```
[the remark this answers is not in the document]
Comment (roduran-dev) [reply apps/workflows/checks.py:88]: But this is from way before this PR. And I didn't change the orignals.

[the remark this answers is not in the document]
Comment (roduran-dev) [reply apps/workflows/checks.py:88]: @JakubFrejlach I think here is another BUG that was here for a few months. (missing values desc2in) 

Created https://redhat.atlassian.net/browse/OSIDB-5025 for the fix

[the remark this answers is not in the document]
Comment (JakubFrejlach) [reply apps/workflows/checks.py:88]: @roduran-dev actually I am not sure whether the `desc2in` should even be in that list in the first place, if you look at the method [`_parse_dict`](https://github.com/RedHatProductSecurity/osidb/blob/993e8bdcd9c91aea26a08db7bb32920027dcd6fb/apps/workflows/checks.py#L95C9-L95C20) `desc2in` and even `desc2not_in` (which is just a negated counterpart for the `desc2in`) are processed there. Also if you look in method [`parse`](https://github.com/RedHatProductSecurity/osidb/blob/993e8bdcd9c91aea26a08db7bb32920027dcd6fb/apps/workflows/checks.py#L66) you can see that there is actually a check which in case of `dict` calls the `_parse_dict` and otherwise `_parse_string`.

File: apps/sla/tests/test_models.py [hunk 16]
```diff
@@ -992,6 +1003,112 @@ def test_in_condition(self, operator, component, accepted):
...
+                impact=Impact.MODERATE,
+            )
+            valid_tracker = TrackerFactory(
+                affects=[valid_affect],
+                embargoed=flaw.embargoed,
+                ps_update_stream=ps_update_stream.name,
+                type=Tracker.BTS2TYPE[ps_module.bts_name],
+            )
+
+            # # Should accept because condition match
```
Comment (JakubFrejlach) [inline apps/sla/tests/test_models.py]: minor nitpick: double hashtag

File: apps/workflows/checks.py [hunk 17]
```diff
@@ -184,24 +191,33 @@ def choices_field(model, name):
...
 
                 return (doc, compare_element)
 
-    def desc2not_equals(self, check_desc):
+    def desc2not_equals(self, check_desc, **kwargs):
         """
         negative attribute to literal value comparison check
 
         currently only supports string attributes
         which values do not contain any spaces
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: stale docstring

File: apps/workflows/checks.py [hunk 18]
```diff
@@ -155,19 +156,25 @@ def has_element(instance):
...
+
+    def desc2equals(self, check_desc, **kwargs):
         """
         attribute to literal value equality check
 
-        currently only supports string attributes
-        which values do not contain any spaces
+        supports string attributes including those with spaces in values
         """
+        original_check_desc = kwargs.get("original_check_desc")
```
Comment (JakubFrejlach) [inline apps/workflows/checks.py]: `original_check_desc` should be an explicit parameter on `desc2equals` and `desc2not_equals` rather than pulled from `**kwargs`, these methods actually depend on it, so the signature should reflect that. The other `desc2*` methods can keep `**kwargs` to absorb it from the dispatch loop.

```python
def desc2equals(self, check_desc, original_check_desc=None, **kwargs):
```

**Review by JakubFrejlach — COMMENTED** [review]
Great improvement, I have a few more things which should be simple to handle.

File: apps/sla/tests/test_models.py (hunk 16 above)
Comment (roduran-dev) [reply apps/sla/tests/test_models.py]: fixed.

File: apps/workflows/checks.py (hunk 18 above)
Comment (roduran-dev) [reply apps/workflows/checks.py]: changed :+1:

**Review by JakubFrejlach — APPROVED** [review]
LGTM now!

File: apps/workflows/checks.py (hunk 17 above)
Comment (Elkasitu) [reply apps/workflows/checks.py]: comment still unaddressed

File: apps/workflows/checks.py [hunk 19]
```diff
@@ -68,6 +68,7 @@ def parse(self, check_desc):
         return self._parse_string(check_desc)
 
     def _parse_string(self, check_desc):
+        original_check_desc = check_desc
         check_desc = check_desc.replace(" ", "_")
```
Comment (Elkasitu) [inline apps/workflows/checks.py:72]: did you consider not converting spaces to underscores in the first place? if we're looking for tokens having spaces or otherwise should not matter

maybe @osoukup knows why we did it like this in the first place?

**Review by Elkasitu — APPROVED** [review]
I think this can be merged into a single commit, have a few nitpicks but LGTM otherwise

**Review by osoukup — APPROVED** [review]
LGTM

File: apps/workflows/checks.py (hunk 19 above)
Comment (osoukup) [reply apps/workflows/checks.py:72]: I think at this point it is mostly historical reasons and if we could get rid of it. I think that originally the aim was to prioritize properties and constants over the free text and those are always underscored. We could however achieve that other ways and this was probably just a simple one. The side effect of not parsing the free text with spaces well was not taken into account.
